dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -261,155 +256,221 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-            PushNotificationTelemetry::ActivatorRegisteredByApi(S_OK);
-        }
+                // Register a COM object for the PushNotificationLongRunningProcess to CoCreate in background activation scenarios
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-        catch(...)
+                scopeExitToCleanRegistrations.release();
+            }
+        }
+        catch (...)
         {
             PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
+
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            registering = false;
+        }
     }
 
     void PushNotificationManager::Unregister()
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(E_FAIL, registering, "Register or Unregister currently in progress!");
+            registering = true;
+        }
+
+        auto scope_exit = wil::scope_exit(
+            [&] {
+                auto lock{ m_lock.lock_exclusive() };
+                registering = false;
+            });
 
         try
         {
-            auto lock{ m_lock.lock_exclusive() };
-            
-            // Check for COM flag, a valid cookie
-            if (AppModel::Identity::IsPackagedProcess())
+            if (PushNotificationHelpers::IsPackagedAppScenario())
             {
-                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_comActivatorRegistration, "ComActivator not registered.");
-                m_comActivatorRegistration.reset();
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_FAIL, !m_comActivatorRegistration, "No active COM registration");
+                    // m_comActivatorRegistration handles multiple unregistrations
+                    m_comActivatorRegistration.reset();
+                }
+
+                auto appUserModelId{ PushNotificationHelpers::GetAppUserModelId() };
+                THROW_IF_FAILED(PushNotifications_UnregisterNotificationSinkForFullTrustApplication(appUserModelId.get()));
+            }
+            else
+            {
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_HR_IF_MSG(E_FAIL, !m_comActivatorRegistration, "No active COM registration");
```

Comment:
Are developers going to see these messages somewhere? If yes, we should try to see something more generic or understandable. The developer will not have any idea of what "COM registration" or "foreground sink" means, especially after having removed the registration flags.

Reply:
Yup. All these error messages will show up for the developer.

Reply:
Changing message to "Need to call Register() before calling Unregister()"